File: sdks/typescript/test/streamed_verify_options.test.ts
// verifyStreamedTurnWithOptions tests — the options-object streamed fast
// path (SPEC §5.13): required_scope against token.granted_scope, single-use
// challenges with the §10 consumption order, and verifier-side session and
// stream binding checks.

import { test } from "node:test";
import assert from "node:assert/strict";

import {
  generateHumanRoot,
  generateAgent,
  issueDelegation,
  issueSessionToken,
  signChallenge,
  generateChallenge,
  verifyBundle,
  verifyStreamedTurnWithOptions,
  MemoryChallengeStore,
  UNKNOWN_CHALLENGE,
  PROTOCOL_VERSION,
  SCOPE_MEETING_ATTEND,
  SCOPE_FILES_READ,
  SCOPE_FILES_WRITE,
  type DelegationCert,
  type HybridPrivateKey,
  type ProofBundle,
  type SessionToken,
  type StreamedTurn,
} from "../src/index.js";

const NOW = () => Math.floor(Date.now() / 1000);
const UNKNOWN_REASON = `unknown_challenge: ${UNKNOWN_CHALLENGE}`;

interface Fixture {
  token: SessionToken;
  secret: Uint8Array;
  agentPriv: HybridPrivateKey;
  now: number;
}

async function fixture(scope: string[]): Promise<Fixture> {
  const { root, privateKey: rootPriv } = await generateHumanRoot();
  const { agent, privateKey: agentPriv } = await generateAgent("Turn Bot", "custom");
  const now = NOW();
  const cert: DelegationCert = {
    cert_id: "turn-cert-001",
    version: PROTOCOL_VERSION,
    issuer_id: root.id,
    issuer_pub_key: root.public_key,
    subject_id: agent.id,
    subject_pub_key: agent.public_key,
    scope,
    constraints: [],
    issued_at: now,
    expires_at: now + 86400,
    signature: { ed25519: new Uint8Array(0), ml_dsa_65: new Uint8Array(0) },
  };
  await issueDelegation(cert, rootPriv);
  const challenge = generateChallenge();
  const sig = await signChallenge(challenge, now, agentPriv);
  const bundle: ProofBundle = {
    agent_id: agent.id,
    agent_pub_key: agent.public_key,
    delegations: [cert],
    challenge,
    challenge_at: now,
    challenge_sig: sig,
  };
  const res = await verifyBundle(bundle, { now });
  assert.equal(res.valid, true, res.error_reason);
  const secret = new Uint8Array(32).fill(0x42);
  const token = issueSessionToken(bundle, res, "session-turn", now, now + 1800, secret);
  return { token, secret, agentPriv, now };
}

async function turnFor(
  f: Fixture,
  challenge: Uint8Array,
  sessionContext?: Uint8Array,
  streamID?: Uint8Array,
  streamSeq?: number,
): Promise<StreamedTurn> {
  const sig = await signChallenge(challenge, f.now, f.agentPriv, sessionContext, streamID, streamSeq);
  return {
    challenge,
    challenge_at: f.now,
    challenge_sig: sig,
    session_context: sessionContext,
    stream_id: streamID,
    stream_seq: streamSeq,
  };
}

test("streamed options: required scope allowed and denied", async () => {
  const f = await fixture([SCOPE_MEETING_ATTEND, SCOPE_FILES_READ]);
  const turn = await turnFor(f, generateChallenge());

  const ok = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    required_scope: SCOPE_MEETING_ATTEND,
    now: f.now,
  });
  assert.equal(ok.valid, true, ok.error_reason);

  const denied = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    required_scope: SCOPE_FILES_WRITE,
    now: f.now,
  });
  assert.equal(denied.valid, false);
  assert.equal(denied.identity_status, "scope_denied");
});

test("streamed options: single-use challenge replay is rejected", async () => {
  const f = await fixture([SCOPE_MEETING_ATTEND]);
  const store = new MemoryChallengeStore(16);
  const { challenge } = await store.issue(undefined, 300);
  const turn = await turnFor(f, challenge);
  const opts = { required_scope: SCOPE_MEETING_ATTEND, challenge_store: store, now: f.now };

  const first = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, opts);
  assert.equal(first.valid, true, first.error_reason);

  const replay = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, opts);
  assert.equal(replay.valid, false);
  assert.equal(replay.error_reason, UNKNOWN_REASON);
});

test("streamed options: forged signature does not consume", async () => {
  const f = await fixture([SCOPE_MEETING_ATTEND]);
  const store = new MemoryChallengeStore(16);
  const { challenge } = await store.issue(undefined, 300);
  const turn = await turnFor(f, challenge);
  const opts = { challenge_store: store, now: f.now };

  const forgedSig = turn.challenge_sig.ed25519.slice();
  forgedSig[0]! ^= 0xff;
  const forged: StreamedTurn = {
    ...turn,
    challenge_sig: { ...turn.challenge_sig, ed25519: forgedSig },
  };
  const res = await verifyStreamedTurnWithOptions(f.token, f.secret, forged, opts);
  assert.equal(res.valid, false);
  assert.match(res.error_reason ?? "", /^bad_challenge_sig/);

  const legit = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, opts);
  assert.equal(legit.valid, true, legit.error_reason);
});

test("streamed options: scope denial still consumes", async () => {
  const f = await fixture([SCOPE_MEETING_ATTEND]);
  const store = new MemoryChallengeStore(16);
  const { challenge } = await store.issue(undefined, 300);
  const turn = await turnFor(f, challenge);

  const denied = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    required_scope: SCOPE_FILES_WRITE,
    challenge_store: store,
    now: f.now,
  });
  assert.equal(denied.identity_status, "scope_denied");

  // The denial happened AFTER consumption: the challenge is spent.
  const retry = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    required_scope: SCOPE_MEETING_ATTEND,
    challenge_store: store,
    now: f.now,
  });
  assert.equal(retry.error_reason, UNKNOWN_REASON);
});

test("streamed options: unknown challenge rejected before crypto", async () => {
  const f = await fixture([SCOPE_MEETING_ATTEND]);
  const store = new MemoryChallengeStore(16);
  const turn = await turnFor(f, generateChallenge());
  const res = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    challenge_store: store,
    now: f.now,
  });
  assert.equal(res.error_reason, UNKNOWN_REASON);
});

test("streamed options: session binding checks", async () => {
  const f = await fixture([SCOPE_MEETING_ATTEND]);
  const ctx = new Uint8Array(32);
  ctx[0] = 7;
  const turn = await turnFor(f, generateChallenge(), ctx);

  const ok = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    session_context: ctx,
    now: f.now,
  });
  assert.equal(ok.valid, true, ok.error_reason);

  const other = new Uint8Array(32);
  other[0] = 8;
  const mismatch = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    session_context: other,
    now: f.now,
  });
  assert.match(mismatch.error_reason ?? "", /^session_context_mismatch/);

  const unverifiable = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, { now: f.now });
  assert.match(unverifiable.error_reason ?? "", /^session_context_unverifiable/);

  const unbound = await turnFor(f, generateChallenge());
  const missing = await verifyStreamedTurnWithOptions(f.token, f.secret, unbound, {
    session_context: ctx,
    now: f.now,
  });
  assert.match(missing.error_reason ?? "", /^missing_session_context/);
});

test("streamed options: stream tracking replay and skip", async () => {
  const f = await fixture([SCOPE_MEETING_ATTEND]);
  const streamID = new Uint8Array(32);
  streamID[0] = 3;
  const turn = await turnFor(f, generateChallenge(), undefined, streamID, 4);

  const ok = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    stream: { stream_id: streamID, last_seen_seq: 3 },
    now: f.now,
  });
  assert.equal(ok.valid, true, ok.error_reason);

  const replay = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    stream: { stream_id: streamID, last_seen_seq: 4 },
    now: f.now,
  });
  assert.match(replay.error_reason ?? "", /^stream_seq_replay/);

  const skip = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    stream: { stream_id: streamID, last_seen_seq: 1 },
    now: f.now,
  });
  assert.match(skip.error_reason ?? "", /^stream_seq_skip/);

  const unverifiable = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, { now: f.now });
  assert.match(unverifiable.error_reason ?? "", /^stream_context_unverifiable/);
});

test("streamed options: stream state is a caller-owned snapshot", async () => {
  // The verifier reads the stream snapshot and never advances it: two
  // distinct valid challenges carrying the same stream_seq BOTH verify
  // against the same snapshot (documented caller-owned semantics). The
  // caller must atomically advance its tracked sequence on success, after
  // which the same-seq turn is rejected as a replay.
  const f = await fixture([SCOPE_MEETING_ATTEND]);
  const streamID = new Uint8Array(32);
  streamID[0] = 5;
  const turn1 = await turnFor(f, generateChallenge(), undefined, streamID, 4);
  const turn2 = await turnFor(f, generateChallenge(), undefined, streamID, 4);
  const snapshot = { stream_id: streamID, last_seen_seq: 3 };

  const res1 = await verifyStreamedTurnWithOptions(f.token, f.secret, turn1, {
    stream: snapshot,
    now: f.now,
  });
  const res2 = await verifyStreamedTurnWithOptions(f.token, f.secret, turn2, {
    stream: snapshot,
    now: f.now,
  });
  assert.equal(res1.valid, true, res1.error_reason);
  assert.equal(res2.valid, true, res2.error_reason);

  const advanced = { stream_id: streamID, last_seen_seq: 4 };
  const replay = await verifyStreamedTurnWithOptions(f.token, f.secret, turn2, {
    stream: advanced,
    now: f.now,
  });
  assert.match(replay.error_reason ?? "", /^stream_seq_replay/);
});

test("streamed options: full-verifier options are rejected, never silently ignored", async () => {
  // TypeScript's structural typing lets a full VerifyOptions variable be
  // assigned where StreamedVerifyOptions is expected, so the boundary
  // must hold at runtime: any options object carrying a
  // full-verifier-only field fails closed with unsupported_option.
  const f = await fixture([SCOPE_MEETING_ATTEND]);
  const turn = await turnFor(f, generateChallenge());

  const fullVerifierOnly: Record<string, unknown>[] = [
    { force_revocation_check: true },
    { revocation: { isRevoked: async () => [false, null] } },
    { is_revoked: () => false },
    { policy: { evaluatePolicy: async () => true } },
    { audit: { logVerification: async () => {} } },
    { context: { requested_amount: 5 } },
    { constraint_evaluators: {} },
    { policy_verdict: {} },
    { policy_secret: new Uint8Array(32) },
    { anchor_resolver: { resolveAnchor: async () => null } },
  ];
  for (const extra of fullVerifierOnly) {
    const res = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
      required_scope: SCOPE_MEETING_ATTEND,
      now: f.now,
      ...extra,
    } as never);
    assert.equal(res.valid, false, `option ${Object.keys(extra)[0]} must be rejected`);
    assert.match(res.error_reason ?? "", /^unsupported_option/);
  }

  // The same call without the foreign field verifies — proving the
  // rejection is about the option, not the turn.
  const ok = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    required_scope: SCOPE_MEETING_ATTEND,
    now: f.now,
  });
  assert.equal(ok.valid, true, ok.error_reason);
});

test("streamed options: token checks still apply", async () => {
  const f = await fixture([SCOPE_MEETING_ATTEND]);
  const turn = await turnFor(f, generateChallenge());

  const badSecret = await verifyStreamedTurnWithOptions(
    f.token,
    new Uint8Array(32).fill(0x99),
    turn,
    { now: f.now },
  );
  assert.match(badSecret.error_reason ?? "", /^session_token_invalid/);

  const expired = await verifyStreamedTurnWithOptions(f.token, f.secret, turn, {
    now: f.now + 31 * 60,
  });
  assert.match(expired.error_reason ?? "", /^session_token_invalid/);
});

File: streamed_verify_options_test.go
package ratify_test

import (
	"strings"
	"testing"
	"time"

	. "github.com/identities-ai/ratify-protocol"
)

// streamedFixture builds a verified session token plus the agent key needed
// to sign per-turn challenges.
type streamedFixture struct {
	token     *SessionToken
	secret    []byte
	agentPriv HybridPrivateKey
	now       time.Time
}

func newStreamedFixture(t *testing.T, scope []string) streamedFixture {
	t.Helper()
	humanRoot, humanPriv, err := GenerateHumanRootKeypair()
	if err != nil {
		t.Fatalf("GenerateHumanRootKeypair: %v", err)
	}
	agent, agentPriv, err := GenerateAgentKeypair("Turn Bot", "custom")
	if err != nil {
		t.Fatalf("GenerateAgentKeypair: %v", err)
	}
	now := time.Now()
	cert := &DelegationCert{
		CertID:        "turn-cert-001",
		Version:       ProtocolVersion,
		IssuerID:      humanRoot.ID,
		IssuerPubKey:  humanRoot.PublicKey,
		SubjectID:     agent.ID,
		SubjectPubKey: agent.PublicKey,
		Scope:         scope,
		IssuedAt:      now.Unix(),
		ExpiresAt:     now.Add(24 * time.Hour).Unix(),
	}
	if err := IssueDelegation(cert, humanPriv); err != nil {
		t.Fatalf("IssueDelegation: %v", err)
	}
	challenge, err := GenerateChallenge()
	if err != nil {
		t.Fatalf("GenerateChallenge: %v", err)
	}
	sig, err := SignChallenge(challenge, now.Unix(), agentPriv)
	if err != nil {
		t.Fatalf("SignChallenge: %v", err)
	}
	bundle := &ProofBundle{
		AgentID:      agent.ID,
		AgentPubKey:  agent.PublicKey,
		Delegations:  []DelegationCert{*cert},
		Challenge:    challenge,
		ChallengeAt:  now.Unix(),
		ChallengeSig: sig,
	}
	res := Verify(bundle, VerifyOptions{Now: now})
	if !res.Valid {
		t.Fatalf("initial Verify: %s — %s", res.IdentityStatus, res.ErrorReason)
	}
	secret := make([]byte, 32)
	for i := range secret {
		secret[i] = 0x42
	}
	token, err := IssueSessionToken(bundle, res, "session-turn", now.Unix(), now.Add(30*time.Minute).Unix(), secret)
	if err != nil {
		t.Fatalf("IssueSessionToken: %v", err)
	}
	return streamedFixture{token: token, secret: secret, agentPriv: agentPriv, now: now}
}

// turnFor signs a fresh turn. sessionContext/streamID/streamSeq ride in the
// signable per §6.4.2.
func (f streamedFixture) turnFor(t *testing.T, challenge []byte, sessionContext, streamID []byte, streamSeq int64) StreamedTurn {
	t.Helper()
	at := f.now.Unix()
	var sig HybridSignature
	var err error
	switch {
	case len(streamID) != 0:
		sig, err = SignChallengeWithStream(challenge, at, sessionContext, streamID, streamSeq, f.agentPriv)
	case len(sessionContext) != 0:
		sig, err = SignChallengeWithSessionContext(challenge, at, sessionContext, f.agentPriv)
	default:
		sig, err = SignChallenge(challenge, at, f.agentPriv)
	}
	if err != nil {
		t.Fatalf("sign turn challenge: %v", err)
	}
	return StreamedTurn{
		Challenge:      challenge,
		ChallengeAt:    at,
		ChallengeSig:   sig,
		SessionContext: sessionContext,
		StreamID:       streamID,
		StreamSeq:      streamSeq,
	}
}

func randomChallenge(t *testing.T) []byte {
	t.Helper()
	c, err := GenerateChallenge()
	if err != nil {
		t.Fatalf("GenerateChallenge: %v", err)
	}
	return c
}

func TestStreamedTurnWithOptionsRequiredScope(t *testing.T) {
	f := newStreamedFixture(t, []string{ScopeMeetingAttend, ScopeFilesRead})

	turn := f.turnFor(t, randomChallenge(t), nil, nil, 0)
	res := VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{
		RequiredScope: ScopeMeetingAttend, Now: f.now,
	})
	if !res.Valid {
		t.Fatalf("granted scope must verify: %s — %s", res.IdentityStatus, res.ErrorReason)
	}

	// A scope the token does not carry is denied — the designed check the
	// positional form could never make.
	res = VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{
		RequiredScope: ScopeFilesWrite, Now: f.now,
	})
	if res.Valid || res.IdentityStatus != IdentityStatusScopeDenied {
		t.Fatalf("got %s / %s, want scope_denied", res.IdentityStatus, res.ErrorReason)
	}
}

func TestStreamedTurnWithOptionsSingleUse(t *testing.T) {
	f := newStreamedFixture(t, []string{ScopeMeetingAttend})
	store := NewMemoryChallengeStore(16)
	challenge, _, err := store.Issue(nil, 5*time.Minute)
	if err != nil {
		t.Fatalf("Issue: %v", err)
	}
	turn := f.turnFor(t, challenge, nil, nil, 0)
	opts := VerifyOptions{RequiredScope: ScopeMeetingAttend, ChallengeStore: store, Now: f.now}

	first := VerifyStreamedTurnWithOptions(f.token, f.secret, turn, opts)
	if !first.Valid {
		t.Fatalf("first presentation: %s — %s", first.IdentityStatus, first.ErrorReason)
	}
	replay := VerifyStreamedTurnWithOptions(f.token, f.secret, turn, opts)
	if replay.Valid || !strings.HasPrefix(replay.ErrorReason, "unknown_challenge") {
		t.Fatalf("replay got %s / %s, want unknown_challenge", replay.IdentityStatus, replay.ErrorReason)
	}
}

func TestStreamedTurnWithOptionsForgeryDoesNotConsume(t *testing.T) {
	f := newStreamedFixture(t, []string{ScopeMeetingAttend})
	store := NewMemoryChallengeStore(16)
	challenge, _, _ := store.Issue(nil, 5*time.Minute)
	turn := f.turnFor(t, challenge, nil, nil, 0)
	opts := VerifyOptions{ChallengeStore: store, Now: f.now}

	forged := turn
	forged.ChallengeSig.Ed25519 = append([]byte(nil), turn.ChallengeSig.Ed25519...)
	forged.ChallengeSig.Ed25519[0] ^= 0xFF
	res := VerifyStreamedTurnWithOptions(f.token, f.secret, forged, opts)
	if res.Valid || !strings.HasPrefix(res.ErrorReason, "bad_challenge_sig") {
		t.Fatalf("forged turn: got %s / %s", res.IdentityStatus, res.ErrorReason)
	}

	// The legitimate presentation still succeeds afterwards.
	if res := VerifyStreamedTurnWithOptions(f.token, f.secret, turn, opts); !res.Valid {
		t.Fatalf("legitimate turn after forgery: %s — %s", res.IdentityStatus, res.ErrorReason)
	}
}

func TestStreamedTurnWithOptionsScopeDenialStillConsumes(t *testing.T) {
	f := newStreamedFixture(t, []string{ScopeMeetingAttend})
	store := NewMemoryChallengeStore(16)
	challenge, _, _ := store.Issue(nil, 5*time.Minute)
	turn := f.turnFor(t, challenge, nil, nil, 0)

	denied := VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{
		RequiredScope: ScopeFilesWrite, ChallengeStore: store, Now: f.now,
	})
	if denied.Valid || denied.IdentityStatus != IdentityStatusScopeDenied {
		t.Fatalf("got %s / %s, want scope_denied", denied.IdentityStatus, denied.ErrorReason)
	}
	// The denial happened AFTER consumption: retrying with the granted
	// scope fails — the challenge is spent.
	retry := VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{
		RequiredScope: ScopeMeetingAttend, ChallengeStore: store, Now: f.now,
	})
	if retry.Valid || !strings.HasPrefix(retry.ErrorReason, "unknown_challenge") {
		t.Fatalf("retry got %s / %s, want unknown_challenge", retry.IdentityStatus, retry.ErrorReason)
	}
}

func TestStreamedTurnWithOptionsUnknownChallengeRejectedBeforeCrypto(t *testing.T) {
	f := newStreamedFixture(t, []string{ScopeMeetingAttend})
	store := NewMemoryChallengeStore(16)
	// Never-issued challenge: rejected with the canonical uniform result.
	turn := f.turnFor(t, randomChallenge(t), nil, nil, 0)
	res := VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{ChallengeStore: store, Now: f.now})
	if res.Valid || !strings.HasPrefix(res.ErrorReason, "unknown_challenge") {
		t.Fatalf("got %s / %s, want unknown_challenge", res.IdentityStatus, res.ErrorReason)
	}
}

func TestStreamedTurnWithOptionsSessionBinding(t *testing.T) {
	f := newStreamedFixture(t, []string{ScopeMeetingAttend})
	ctx := make([]byte, 32)
	ctx[0] = 7

	// Bound turn against the matching verifier context verifies.
	turn := f.turnFor(t, randomChallenge(t), ctx, nil, 0)
	res := VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{SessionContext: ctx, Now: f.now})
	if !res.Valid {
		t.Fatalf("bound turn: %s — %s", res.IdentityStatus, res.ErrorReason)
	}

	// Wrong verifier context is a mismatch.
	other := make([]byte, 32)
	other[0] = 8
	res = VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{SessionContext: other, Now: f.now})
	if res.Valid || !strings.HasPrefix(res.ErrorReason, "session_context_mismatch") {
		t.Fatalf("got %s / %s, want session_context_mismatch", res.IdentityStatus, res.ErrorReason)
	}

	// Presented binding with no verifier expectation is unverifiable.
	res = VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{Now: f.now})
	if res.Valid || !strings.HasPrefix(res.ErrorReason, "session_context_unverifiable") {
		t.Fatalf("got %s / %s, want session_context_unverifiable", res.IdentityStatus, res.ErrorReason)
	}

	// Verifier expectation with an unbound turn is missing.
	unbound := f.turnFor(t, randomChallenge(t), nil, nil, 0)
	res = VerifyStreamedTurnWithOptions(f.token, f.secret, unbound, VerifyOptions{SessionContext: ctx, Now: f.now})
	if res.Valid || !strings.HasPrefix(res.ErrorReason, "missing_session_context") {
		t.Fatalf("got %s / %s, want missing_session_context", res.IdentityStatus, res.ErrorReason)
	}
}

func TestStreamedTurnWithOptionsStreamTracking(t *testing.T) {
	f := newStreamedFixture(t, []string{ScopeMeetingAttend})
	streamID := make([]byte, 32)
	streamID[0] = 3

	// Next-in-sequence turn verifies.
	turn := f.turnFor(t, randomChallenge(t), nil, streamID, 4)
	res := VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{
		Stream: &StreamContext{StreamID: streamID, LastSeenSeq: 3}, Now: f.now,
	})
	if !res.Valid {
		t.Fatalf("in-sequence turn: %s — %s", res.IdentityStatus, res.ErrorReason)
	}

	// Replayed sequence number is rejected.
	res = VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{
		Stream: &StreamContext{StreamID: streamID, LastSeenSeq: 4}, Now: f.now,
	})
	if res.Valid || !strings.HasPrefix(res.ErrorReason, "stream_seq_replay") {
		t.Fatalf("got %s / %s, want stream_seq_replay", res.IdentityStatus, res.ErrorReason)
	}

	// Skipped sequence number is rejected.
	res = VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{
		Stream: &StreamContext{StreamID: streamID, LastSeenSeq: 1}, Now: f.now,
	})
	if res.Valid || !strings.HasPrefix(res.ErrorReason, "stream_seq_skip") {
		t.Fatalf("got %s / %s, want stream_seq_skip", res.IdentityStatus, res.ErrorReason)
	}

	// Stream-bound turn with no verifier stream context is unverifiable.
	res = VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{Now: f.now})
	if res.Valid || !strings.HasPrefix(res.ErrorReason, "stream_context_unverifiable") {
		t.Fatalf("got %s / %s, want stream_context_unverifiable", res.IdentityStatus, res.ErrorReason)
	}
}

func TestStreamedTurnWithOptionsTokenChecksStillApply(t *testing.T) {
	f := newStreamedFixture(t, []string{ScopeMeetingAttend})
	turn := f.turnFor(t, randomChallenge(t), nil, nil, 0)

	// Nil token.
	res := VerifyStreamedTurnWithOptions(nil, f.secret, turn, VerifyOptions{Now: f.now})
	if res.Valid || !strings.HasPrefix(res.ErrorReason, "nil_session_token") {
		t.Fatalf("got %s / %s, want nil_session_token", res.IdentityStatus, res.ErrorReason)
	}

	// Wrong secret fails the HMAC.
	res = VerifyStreamedTurnWithOptions(f.token, []byte("wrong-secret-wrong-secret-wrong!"), turn, VerifyOptions{Now: f.now})
	if res.Valid || !strings.HasPrefix(res.ErrorReason, "session_token_invalid") {
		t.Fatalf("got %s / %s, want session_token_invalid", res.IdentityStatus, res.ErrorReason)
	}

	// Expired token window.
	res = VerifyStreamedTurnWithOptions(f.token, f.secret, turn, VerifyOptions{Now: f.now.Add(31 * time.Minute)})
	if res.Valid || !strings.HasPrefix(res.ErrorReason, "session_token_invalid") {
		t.Fatalf("got %s / %s, want session_token_invalid", res.IdentityStatus, res.ErrorReason)
	}
}
